{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[leiden algo documentation](https://leidenalg.readthedocs.io/en/latest/install.html)\n",
    "\n",
    "# General Approach\n",
    "\n",
    "In this analysis, we explore the clustering of scientific publications using the Leiden algorithm, with a focus on adjusting the algorithm's parameters to improve the coherence and relevance of the clusters formed. The Leiden algorithm, which improves upon the limitations of modularity optimization seen in previous methods, is employed for its effectiveness in identifying communities within networks. We particularly experiment with various resolution parameters, iterating through a range of values to tune the granularity of the clustering results. The objective is to find an optimal balance where clusters are neither too broad, encompassing multiple unrelated topics, nor too narrow, breaking down coherent subjects into excessive subgroups.\n",
    "\n",
    "The analysis starts with the application of the Leiden algorithm using default settings and then progresses to more sophisticated implementations, including adjusting the resolution parameter to influence the size and coherence of clusters. This parameter tuning is crucial for addressing the resolution limit problem inherent in modularity-based community detection methods. By carefully selecting resolution parameters, we aim to achieve a set of clusters that are meaningful and manageable for qualitative analysis.\n",
    "\n",
    "This process is supported by the creation of summary sheets and detailed exploratory sheets for each parameter set, facilitating a thorough review of the clustering outcomes. Parameters leading to the most coherent and distinct clusters, based on predefined decision criteria, are identified and selected for further analysis. The final selection of parameters is based on qualitative assessments of cluster coherence, ensuring that the clusters formed are both scientifically relevant and insightful for further research explorations.\n",
    "\n",
    "Notes about the leiden algorithm:\n",
    "\n",
    "### The most simpe implementation is this:\n",
    "\n",
    "Here we would just use all default parameters.\n",
    "The metrices are good, the qualitative assessment poor.\n",
    "The `ModularityVertexPartition` is the default partitioning algorithm\n",
    "\n",
    "- it is based on the modularity measure (link density within communities vs. link density between communities)\n",
    "- Higher scires = stronger division into communities\n",
    "- It tries to maximize the modularity score\n",
    "\n",
    "The problem here is that the partitions are really big (if run without max_comm_size, around 5,000 papers in the largest communit)\n",
    "Examining the most central articles (based on the eigenvector similarity), they do not seem coherent.\n",
    "\n",
    "**PROBLEM:** Modularity, though\n",
    "robust for many practical applications, suffers from the resolution limit problem,\n",
    "in which optimization may fail to identify clusters smaller than a certain scale\n",
    "that is dependent on properties of the network. (https://arxiv.org/pdf/2308.09644.pdf)\n",
    "\n",
    "### The second implementation is this:\n",
    "\n",
    "We are trying to tune the resolution parameter. To access it, we use the CPMVertexPartition algorithm. (CPM = Constant Potts Model)\n",
    "This means that modularity is not an important parameter anymore.\n",
    "\n",
    "- The resolution parameter is the only parameter of this algorithm\n",
    "- It compares the actual link density to the resolution parameter\n",
    "- If the link density is higher than the resolution parameter, the nodes are put into the same community\n",
    "- It has a quality attribute, which is similar to the modularity score\n",
    "- We will not rely on this but use a qualitative analysis to assess the results\n",
    "\n",
    "In previous papers (Waltman 2020a, Ahlgren 2020), the authors used different levels of granularity. They used 10 different values of the resolution parameter γ: 0.000001, 0.000002, 0.000005, 0.00001, 0.00002, 0.00005, 0.0001, 0.0002, 0.0005, 0.001, 0.002, 0.005, 0.01\n",
    "\n",
    "We decide to use the following resolution parameter to narrow down our search.\n",
    "0.00001,0.00002,0.0001,0.0002,0.001,0.002,0.01,0.02,\n",
    "\n",
    "We then evaluated the solutions\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Packages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import leidenalg as la\n",
    "import igraph as ig\n",
    "import sys\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "\n",
    "sys.path.append(\"/Users/jlq293/Projects/Study-1-Bibliometrics/src/network/\")\n",
    "from PartitionCreator import PartitionCreator\n",
    "from NetworkAnalyzer import TextAnalyzer, CommunityExplorer, ClusterPlotter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['alpha0.3_k10', 'alpha0.3_k15', 'alpha0.3_k20', 'alpha0.3_k5', 'alpha0.5_k10', 'alpha0.5_k15', 'alpha0.5_k20', 'alpha0.5_k5'])\n"
     ]
    }
   ],
   "source": [
    "def load_graph_files():\n",
    "    \"\"\"\n",
    "    Load graph files and return a dictionary of graphs and a DataFrame.\n",
    "\n",
    "    Returns:\n",
    "        params_graph_dict (dict): A dictionary containing graph objects, where the keys are the graph names.\n",
    "        df (DataFrame): A DataFrame containing embeddings.\n",
    "\n",
    "    \"\"\"\n",
    "    graph_files = glob.glob(\"../data/05-graphs/weighted-knn-citation-graph/*.graphml\")\n",
    "    graph_files.sort()\n",
    "\n",
    "    params_graph_dict = {\n",
    "        f.split(\"/\")[-1]\n",
    "        .replace(\"weighted_\", \"\")\n",
    "        .replace(\"_knn_citation.graphml\", \"\"): ig.Graph.Read_GraphML(f)\n",
    "        for f in graph_files\n",
    "    }\n",
    "\n",
    "    df = pd.read_pickle(\"../data/04-embeddings/df_with_specter2_embeddings.pkl\")\n",
    "\n",
    "    return params_graph_dict, df\n",
    "\n",
    "\n",
    "params_graph_dict, df = load_graph_files()\n",
    "print(params_graph_dict.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ModularityVertexPartition\n",
    "\n",
    "not favorable\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>params</th>\n",
       "      <th>modularity</th>\n",
       "      <th>nr_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.683545</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>alpha0.5_k20</td>\n",
       "      <td>0.682633</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.679827</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.679479</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alpha0.3_k20</td>\n",
       "      <td>0.679333</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alpha0.3_k10</td>\n",
       "      <td>0.676278</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.668973</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>alpha0.3_k5</td>\n",
       "      <td>0.665366</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         params  modularity  nr_clusters\n",
       "5  alpha0.5_k15    0.683545           29\n",
       "6  alpha0.5_k20    0.682633           29\n",
       "4  alpha0.5_k10    0.679827           29\n",
       "1  alpha0.3_k15    0.679479           34\n",
       "2  alpha0.3_k20    0.679333           32\n",
       "0  alpha0.3_k10    0.676278           32\n",
       "7   alpha0.5_k5    0.668973           30\n",
       "3   alpha0.3_k5    0.665366           36"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modularity = []\n",
    "nr_clusters = []\n",
    "params = []\n",
    "for params, G in params_graph_dict.items():\n",
    "    pc = PartitionCreator(G, df)\n",
    "    pc.create_partition_from_modularityvertexpartition(\n",
    "        n_iterations=20,\n",
    "        # max_comm_size=1500\n",
    "    )\n",
    "    modularity.append(pc.modularity)\n",
    "    nr_clusters.append(len(pc.cluster_sizes))\n",
    "\n",
    "# create df\n",
    "df_params = pd.DataFrame(\n",
    "    {\n",
    "        \"params\": list(params_graph_dict.keys()),\n",
    "        \"modularity\": modularity,\n",
    "        \"nr_clusters\": nr_clusters,\n",
    "    }\n",
    ").sort_values(by=\"modularity\", ascending=False)\n",
    "df_params"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CPM RESOLUTION DETERMINATION\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# waltman 2020a:\n",
    "\n",
    "Different levels of granularity were considered. For each relatedness measure, we obtained 10 clustering solutions, each of them for a different value of the resolution parameter γ. The following values of γ were used: 0.00001, 0.00002, 0.00005, 0.0001, 0.0002, 0.0005, 0.001, 0.002, 0.005, and 0.01.\n",
    "\n",
    "# ahlgren 2020:\n",
    "\n",
    "Using different values of the resolution parameter γ (0.000001, 0.000002, 0.000005, 0.00001, 0.00002, 0.00005, 0.0001, 0.0002, 0.0005, 0.001, 0.002), we obtain 11 clustering solutions for each relatedness measure. Compared to our earlier study (Ahlgren et al., 2019), we exclude the clustering solutions for the two largest resolution values used in that study (0.005 and 0.01). These clustering solutions have around 300,000 and 500,000 clusters, respectively, and most of the clusters consist of fewer than 10 publications.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_params_graph_dict(params_graph_dict):\n",
    "    \"\"\"\n",
    "    Process the params_graph_dict to create partitions and explore communities.\n",
    "    Args:\n",
    "        params_graph_dict (dict): A dictionary containing graph objects, where the keys are the graph names.\n",
    "    Returns:\n",
    "        partitions_explorer_dict (dict): A dictionary containing graph summary dataframes.\n",
    "        nr_clusters (list): A list of the number of clusters for each graph.\n",
    "        graph_params (list): A list of the graph parameters.\n",
    "        resolution_params (list): A list of the resolution parameters.\n",
    "\n",
    "    \"\"\"\n",
    "    # Define the resolution values to iterate over\n",
    "    resolution_values = [\n",
    "        0.000001,\n",
    "        0.000002,\n",
    "        0.000005,\n",
    "        0.00001,\n",
    "        0.00002,\n",
    "        0.00005,\n",
    "        0.0001,\n",
    "        0.0002,\n",
    "        0.0005,\n",
    "        0.001,\n",
    "        0.002,\n",
    "        0.006,\n",
    "        0.005,\n",
    "        0.01,\n",
    "        0.02,\n",
    "        0.05,\n",
    "    ]\n",
    "    iterations = 20\n",
    "    nr_clusters = []\n",
    "    graph_params = []\n",
    "    resolution_params = []\n",
    "\n",
    "    partitions_explorer_dict = {}\n",
    "\n",
    "    # Iterate over each graph and resolution value\n",
    "    for params, G in params_graph_dict.items():\n",
    "        for resolution in resolution_values:\n",
    "            # Create partition using CPMVertexPartition algorithm\n",
    "            pc = PartitionCreator(G, df)\n",
    "            pc.create_partition_from_cmpvertexpartition(\n",
    "                n_iterations=iterations,\n",
    "                resolution_parameter=resolution,\n",
    "                verbose=False,\n",
    "                cluster_colunm_name=f\"cluster_{params}_res{resolution}\",\n",
    "                centrality_column_name=None,\n",
    "            )\n",
    "\n",
    "            # Explore communities using CommunityExplorer\n",
    "            ce = CommunityExplorer(\n",
    "                pc.df, cluster_column=f\"cluster_{params}_res{resolution}\"\n",
    "            )\n",
    "            ce.create_cluster_sheets(sort=False, n=15)\n",
    "            ce.create_summary_sheet()\n",
    "            df_summary, _ = ce.full_return()\n",
    "            partitions_explorer_dict[f\"{params}_res{resolution}\"] = df_summary\n",
    "            nr_clusters.append(len(pc.partition.sizes()))\n",
    "            graph_params.append(params)\n",
    "            resolution_params.append(resolution)\n",
    "\n",
    "    # Create a DataFrame with the graph parameters and the number of clusters\n",
    "    df_graph_cluster_params = pd.DataFrame(\n",
    "        {\n",
    "            \"params\": graph_params,\n",
    "            \"resolution\": resolution_params,\n",
    "            \"nr_clusters\": nr_clusters,\n",
    "        }\n",
    "    ).sort_values(by=\"nr_clusters\", ascending=False)\n",
    "\n",
    "    return partitions_explorer_dict, df_graph_cluster_params\n",
    "\n",
    "\n",
    "partitions_explorer_dict, df_graph_cluster_params = process_params_graph_dict(\n",
    "    params_graph_dict\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>params</th>\n",
       "      <th>resolution</th>\n",
       "      <th>nr_clusters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.01000</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.01000</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>alpha0.5_k20</td>\n",
       "      <td>0.01000</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>alpha0.3_k10</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>alpha0.3_k10</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>alpha0.3_k20</td>\n",
       "      <td>0.01000</td>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>alpha0.3_k5</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>alpha0.5_k20</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>alpha0.3_k20</td>\n",
       "      <td>0.00600</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>alpha0.5_k20</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>alpha0.3_k5</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>alpha0.3_k20</td>\n",
       "      <td>0.00500</td>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>alpha0.3_k10</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00005</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>alpha0.3_k5</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00020</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>alpha0.5_k20</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>alpha0.3_k10</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>alpha0.3_k20</td>\n",
       "      <td>0.00200</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>alpha0.5_k15</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>alpha0.3_k15</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>alpha0.5_k5</td>\n",
       "      <td>0.00010</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>alpha0.5_k10</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           params  resolution  nr_clusters\n",
       "93   alpha0.5_k15     0.01000          477\n",
       "75   alpha0.5_k10     0.00600          442\n",
       "122   alpha0.5_k5     0.00200          433\n",
       "29   alpha0.3_k15     0.01000          379\n",
       "109  alpha0.5_k20     0.01000          375\n",
       "76   alpha0.5_k10     0.00500          373\n",
       "11   alpha0.3_k10     0.00600          362\n",
       "12   alpha0.3_k10     0.00500          320\n",
       "45   alpha0.3_k20     0.01000          299\n",
       "91   alpha0.5_k15     0.00600          298\n",
       "58    alpha0.3_k5     0.00200          297\n",
       "121   alpha0.5_k5     0.00100          272\n",
       "92   alpha0.5_k15     0.00500          256\n",
       "27   alpha0.3_k15     0.00600          241\n",
       "107  alpha0.5_k20     0.00600          231\n",
       "28   alpha0.3_k15     0.00500          218\n",
       "43   alpha0.3_k20     0.00600          190\n",
       "108  alpha0.5_k20     0.00500          187\n",
       "74   alpha0.5_k10     0.00200          173\n",
       "57    alpha0.3_k5     0.00100          165\n",
       "44   alpha0.3_k20     0.00500          158\n",
       "10   alpha0.3_k10     0.00200          155\n",
       "120   alpha0.5_k5     0.00050          152\n",
       "90   alpha0.5_k15     0.00200          111\n",
       "26   alpha0.3_k15     0.00200          104\n",
       "117   alpha0.5_k5     0.00005          104\n",
       "56    alpha0.3_k5     0.00050           94\n",
       "73   alpha0.5_k10     0.00100           93\n",
       "119   alpha0.5_k5     0.00020           93\n",
       "106  alpha0.5_k20     0.00200           89\n",
       "9    alpha0.3_k10     0.00100           86\n",
       "42   alpha0.3_k20     0.00200           79\n",
       "89   alpha0.5_k15     0.00100           63\n",
       "25   alpha0.3_k15     0.00100           60\n",
       "118   alpha0.5_k5     0.00010           55\n",
       "72   alpha0.5_k10     0.00050           54"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only cluster nr betwe 50 and 500\n",
    "df_graph_cluster_params_lim = df_graph_cluster_params[\n",
    "    (df_graph_cluster_params[\"nr_clusters\"] > 50)\n",
    "    & (df_graph_cluster_params[\"nr_clusters\"] < 500)\n",
    "]\n",
    "df_graph_cluster_params_lim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_graph_summary_to_excel(partitions_explorer_dict, filename):\n",
    "    \"\"\"\n",
    "    Save partitions_explorer_dict to an Excel file, with each value as a separate sheet.\n",
    "\n",
    "    Args:\n",
    "        partitions_explorer_dict (dict): A dictionary containing graph summary dataframes.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    with pd.ExcelWriter(f\"../output/tables/cluster-explorer/{filename}\") as writer:\n",
    "        for k, v in list(partitions_explorer_dict.items()):\n",
    "            # if \"_k5_\" in k:\n",
    "            #    continue\n",
    "            v.to_excel(writer, sheet_name=k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_graph_summary_to_excel(\n",
    "    df_graph_cluster_params_lim, filename=\"FullParameterFinder_ClusterExplorer.xlsx\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Limit Parameter Space\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decision Criteria:\n",
    "\n",
    "1. premature ejaculation and sexual dysfunction be distinct clusters\n",
    "2. Mixing of anxiety and panic disorder?\n",
    "3. Diabetic Mice VS Body Weight and Diabetes VS depression treatment in diabetics?\n",
    "4. Use alpha = 0.3 (look at alpha0.5_k10_res0.002); topics: 50, 133 -> very mixed words, little coherence.Not the case for alpha0.3_k10_res0.002 (here, sexual dys and premature ejac single topic though)\n",
    "5. Zimeldine (1 or two topics)\n",
    "6. Pregnancy related topics.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "keepornot = {\n",
    "    # all k=5 have many many clusters of single papers\n",
    "    \"alpha0.5_k5_res1e-05\": 0,\n",
    "    \"alpha0.5_k5_res2e-05\": 0,\n",
    "    \"alpha0.5_k5_res0.0001\": 0,\n",
    "    \"alpha0.5_k5_res0.0002\": 0,\n",
    "    \"alpha0.5_k5_res0.001\": 0,\n",
    "    \"alpha0.5_k5_res0.002\": 0,\n",
    "    \"alpha0.5_k5_res0.006\": 0,\n",
    "    \"alpha0.5_k5_res0.01\": 0,\n",
    "    \"alpha0.5_k5_res0.02\": 0,\n",
    "    # K10\n",
    "    \"alpha0.5_k10_res1e-05\": 0,\n",
    "    \"alpha0.5_k10_res2e-05\": 0,\n",
    "    \"alpha0.5_k10_res0.0001\": 0,\n",
    "    \"alpha0.5_k10_res0.0002\": 0,\n",
    "    \"alpha0.5_k10_res0.001\": 0,  # premature ejaculation and sexual dysfunction and one topic\n",
    "    \"alpha0.5_k10_res0.002\": 0,  # weight and sex are one topic, same problem as above\n",
    "    \"alpha0.5_k10_res0.006\": 0,\n",
    "    \"alpha0.5_k10_res0.01\": 0,\n",
    "    \"alpha0.5_k10_res0.02\": 0,\n",
    "    # K15\n",
    "    \"alpha0.5_k15_res1e-05\": 0,\n",
    "    \"alpha0.5_k15_res2e-05\": 0,\n",
    "    \"alpha0.5_k15_res0.0001\": 0,\n",
    "    \"alpha0.5_k15_res0.0002\": 0,\n",
    "    \"alpha0.5_k15_res0.001\": 0,  # no bad, very broad and inclusive topics\n",
    "    \"alpha0.5_k15_res0.002\": 0,  # gut sex dysf and premature ejaculation are one topic\n",
    "    \"alpha0.5_k15_res0.006\": 1,  # pretty good. bit too many. too many pregnancy clusters, for example\n",
    "    \"alpha0.5_k15_res0.01\": 0,\n",
    "    \"alpha0.5_k15_res0.02\": 0,\n",
    "    # K20\n",
    "    \"alpha0.5_k20_res1e-05\": 0,\n",
    "    \"alpha0.5_k20_res2e-05\": 0,\n",
    "    \"alpha0.5_k20_res0.0001\": 0,\n",
    "    \"alpha0.5_k20_res0.0002\": 0,\n",
    "    \"alpha0.5_k20_res0.001\": 0,\n",
    "    \"alpha0.5_k20_res0.002\": 0,\n",
    "    \"alpha0.5_k20_res0.006\": 1,\n",
    "    \"alpha0.5_k20_res0.01\": 1,  # MANY MANT HT TOPCIS (something betwe 002 and 01 needed.)\n",
    "    \"alpha0.5_k20_res0.02\": 0,\n",
    "    # NEW ALPHA 0.3\n",
    "    # K5\n",
    "    \"alpha0.3_k5_res1e-05\": 0,\n",
    "    \"alpha0.3_k5_res2e-05\": 0,\n",
    "    \"alpha0.3_k5_res0.0001\": 0,\n",
    "    \"alpha0.3_k5_res0.0002\": 0,\n",
    "    \"alpha0.3_k5_res0.001\": 0,\n",
    "    \"alpha0.3_k5_res0.002\": 0,\n",
    "    \"alpha0.3_k5_res0.006\": 0,\n",
    "    \"alpha0.3_k5_res0.01\": 0,\n",
    "    \"alpha0.3_k5_res0.02\": 0,\n",
    "    # K10\n",
    "    \"alpha0.3_k10_res1e-05\": 0,\n",
    "    \"alpha0.3_k10_res2e-05\": 0,\n",
    "    \"alpha0.3_k10_res0.0001\": 0,\n",
    "    \"alpha0.3_k10_res0.0002\": 0,\n",
    "    \"alpha0.3_k10_res0.001\": 0,  # premature ejaculation and sexual dysfunction and one topic; anxiety + panic too\n",
    "    \"alpha0.3_k10_res0.002\": 1,  # gut\n",
    "    \"alpha0.3_k10_res0.006\": 1,\n",
    "    \"alpha0.3_k10_res0.01\": 0,  # good illustration for granularity using 'suicide' (500 topics)\n",
    "    \"alpha0.3_k10_res0.02\": 0,\n",
    "    # K15\n",
    "    \"alpha0.3_k15_res1e-05\": 0,\n",
    "    \"alpha0.3_k15_res2e-05\": 0,\n",
    "    \"alpha0.3_k15_res0.0001\": 0,\n",
    "    \"alpha0.3_k15_res0.0002\": 0,\n",
    "    \"alpha0.3_k15_res0.001\": 0,  # no bad, very broad and inclusive topics\n",
    "    \"alpha0.3_k15_res0.002\": 0,  # gut sex dysf and premature ejaculation are one topic\n",
    "    \"alpha0.3_k15_res0.006\": 1,  # pretty good. bit too many. too many pregnancy clusters, for example\n",
    "    \"alpha0.3_k15_res0.01\": 1,  # good illustration for granularity using 'suicide' (500 topics)\n",
    "    \"alpha0.3_k15_res0.02\": 0,\n",
    "    # K 20\n",
    "    \"alpha0.3_k20_res1e-05\": 0,\n",
    "    \"alpha0.3_k20_res2e-05\": 0,\n",
    "    \"alpha0.3_k20_res0.0001\": 0,\n",
    "    \"alpha0.3_k20_res0.0002\": 0,\n",
    "    \"alpha0.3_k20_res0.001\": 0,\n",
    "    \"alpha0.3_k20_res0.002\": 0,  # too few?\n",
    "    \"alpha0.3_k20_res0.006\": 1,  # best ?\n",
    "    \"alpha0.3_k20_res0.01\": 1,  # too many? aber gut.\n",
    "    \"alpha0.3_k20_res0.02\": 1,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### save limit param space to excel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>params</th>\n",
       "      <th>cl_nr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alpha0.3_k10_res0.002</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alpha0.3_k20_res0.006</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>alpha0.5_k20_res0.006</td>\n",
       "      <td>231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alpha0.3_k15_res0.006</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>alpha0.5_k15_res0.006</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>alpha0.3_k20_res0.01</td>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alpha0.3_k10_res0.006</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>alpha0.5_k20_res0.01</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>alpha0.3_k15_res0.01</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>alpha0.3_k20_res0.02</td>\n",
       "      <td>563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  params  cl_nr\n",
       "0  alpha0.3_k10_res0.002    155\n",
       "4  alpha0.3_k20_res0.006    190\n",
       "8  alpha0.5_k20_res0.006    231\n",
       "2  alpha0.3_k15_res0.006    241\n",
       "7  alpha0.5_k15_res0.006    298\n",
       "5   alpha0.3_k20_res0.01    299\n",
       "1  alpha0.3_k10_res0.006    362\n",
       "9   alpha0.5_k20_res0.01    375\n",
       "3   alpha0.3_k15_res0.01    379\n",
       "6   alpha0.3_k20_res0.02    563"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "limited_parameter_data_dict = {\n",
    "    k: v for k, v in partitions_explorer_dict.items() if keepornot.get(k, 0) == 1\n",
    "}\n",
    "\n",
    "save_graph_summary_to_excel(\n",
    "    limited_parameter_data_dict, filename=\"LimitedParameterFinder_ClusterExplorer.xlsx\"\n",
    ")\n",
    "\n",
    "# show summary\n",
    "df_lim_param = pd.DataFrame(\n",
    "    {\"cl_nr\": {k: len(v) for k, v in limited_parameter_data_dict.items()}}\n",
    ")\n",
    "\n",
    "# Reset the index to turn the keys into a column and then sort by 'cl_nr'\n",
    "df_lim_param = (\n",
    "    df_lim_param.reset_index()\n",
    "    .rename(columns={\"index\": \"params\"})\n",
    "    .sort_values(by=\"cl_nr\")\n",
    ")\n",
    "df_lim_param"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_to_excel_with_hyperlinks(filepath, df_summary, comb_sheets):\n",
    "    \"\"\"\n",
    "    Saves summary DataFrame and combined sheets to an Excel file,\n",
    "    with hyperlinks from the combined sheets back to the summary sheet,\n",
    "    and from the summary sheet to each combined sheet.\n",
    "\n",
    "    Parameters:\n",
    "    - filepath: The path to the Excel file to save.\n",
    "    - df_summary: The summary DataFrame to save to the 'summary' sheet.\n",
    "    - comb_sheets: A list of DataFrames to save to individual sheets, with hyperlinks back to the summary.\n",
    "    \"\"\"\n",
    "    with pd.ExcelWriter(filepath, engine=\"openpyxl\") as writer:\n",
    "        # Write the summary sheet\n",
    "        df_summary.to_excel(writer, sheet_name=\"summary\", index=False)\n",
    "\n",
    "        # Access the workbook through the writer.book attribute after writing\n",
    "        workbook = writer.book\n",
    "\n",
    "        # Iterate through comb_sheets and write each to a new sheet\n",
    "        for i, s in enumerate(comb_sheets):\n",
    "            sheet_name = f\"cluster_{i}\"\n",
    "            s.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "\n",
    "            # Access the current sheet\n",
    "            current_sheet = workbook[sheet_name]\n",
    "\n",
    "            # find max row and add 5\n",
    "            lastrow = current_sheet.max_row\n",
    "\n",
    "            # Add a hyperlink in A20 to go back to the summary sheet\n",
    "            current_sheet.cell(row=lastrow + 7, column=1).hyperlink = \"#summary!A1\"\n",
    "            current_sheet.cell(row=lastrow + 7, column=1).value = \"Back to Summary\"\n",
    "            current_sheet.cell(row=lastrow + 7, column=1).style = \"Hyperlink\"\n",
    "            # Add nr pubs info\n",
    "            current_sheet.cell(\n",
    "                row=lastrow + 5, column=1\n",
    "            ).value = f\"NrPubs {df_summary.iloc[i, 1]}\"\n",
    "            # add time info\n",
    "            current_sheet.cell(\n",
    "                row=lastrow + 3, column=1\n",
    "            ).value = f\"Year(x,sd) m={df_summary.iloc[i, 2]} sd={df_summary.iloc[i, 3]}\"\n",
    "\n",
    "        # Update the summary sheet with hyperlinks to each cluster sheet\n",
    "        summary_sheet = workbook[\"summary\"]\n",
    "        for row in range(2, summary_sheet.max_row + 1):\n",
    "            cluster_num = summary_sheet.cell(row=row, column=1).value\n",
    "            summary_sheet.cell(\n",
    "                row=row, column=1\n",
    "            ).hyperlink = f\"#'cluster_{cluster_num}'!A1\"\n",
    "            summary_sheet.cell(row=row, column=1).style = \"Hyperlink\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Params: alpha0.3_k10_res0.002; Nr of clusters: 155\n",
      "##################################################\n",
      "Params: alpha0.3_k10_res0.006; Nr of clusters: 362\n",
      "##################################################\n",
      "Params: alpha0.3_k15_res0.006; Nr of clusters: 241\n",
      "##################################################\n",
      "Params: alpha0.3_k15_res0.01; Nr of clusters: 379\n",
      "##################################################\n",
      "Params: alpha0.3_k20_res0.006; Nr of clusters: 190\n",
      "##################################################\n",
      "Params: alpha0.3_k20_res0.01; Nr of clusters: 299\n",
      "##################################################\n",
      "Params: alpha0.3_k20_res0.02; Nr of clusters: 563\n",
      "##################################################\n",
      "Params: alpha0.5_k15_res0.006; Nr of clusters: 298\n",
      "##################################################\n",
      "Params: alpha0.5_k20_res0.006; Nr of clusters: 231\n",
      "##################################################\n",
      "Params: alpha0.5_k20_res0.01; Nr of clusters: 375\n",
      "##################################################\n"
     ]
    }
   ],
   "source": [
    "iterations = 20\n",
    "quality_values = []\n",
    "nr_clusters = []\n",
    "partitions = []\n",
    "graph_summary_df_dict = {}\n",
    "\n",
    "for params in limited_parameter_data_dict.keys():\n",
    "    G = params_graph_dict[params.split(\"_res\")[0]]\n",
    "    resolution = pd.to_numeric(params.split(\"res\")[1])\n",
    "    column_name = f\"cluster_{params}\"\n",
    "    pc = PartitionCreator(G, df)\n",
    "    pc.create_partition_from_cmpvertexpartition(\n",
    "        n_iterations=iterations,\n",
    "        resolution_parameter=resolution,\n",
    "        verbose=False,\n",
    "        cluster_colunm_name=column_name,\n",
    "        centrality_column_name=f\"centrality_{params}\",\n",
    "    )\n",
    "    #############################\n",
    "    #############################\n",
    "    ce = CommunityExplorer(pc.df, f\"cluster_{params}\")\n",
    "    ce.create_cluster_sheets(sort=True, n=10, sort_column=f\"centrality_{params}\")\n",
    "    sorted_sheets = [s.reset_index(drop=True) for s in ce.sheets]\n",
    "    # suffix\n",
    "    sorted_sheets = [s.add_suffix(f\"_sorted\") for s in sorted_sheets]\n",
    "    ce.create_cluster_sheets(sort=False, n=10, sort_column=f\"centrality_{params}\")\n",
    "    unsorted_sheets = [s.reset_index(drop=True) for s in ce.sheets]\n",
    "    # suffix\n",
    "    unsorted_sheets = [s.add_suffix(f\"_random\") for s in unsorted_sheets]\n",
    "    comb_sheets = [\n",
    "        pd.concat([sorted_sheets[i], unsorted_sheets[i]], axis=1)\n",
    "        for i in range(len(sorted_sheets))\n",
    "    ]\n",
    "    ce.create_summary_sheet()\n",
    "    df_summary = ce.df_summary\n",
    "    graph_summary_df_dict[f\"{params}\"] = [df_summary, comb_sheets]\n",
    "    # save to excel\n",
    "    filepath = f\"../output/tables/cluster-explorer/SingleSolExplorer_{params}.xlsx\"\n",
    "    save_to_excel_with_hyperlinks(filepath, df_summary, comb_sheets)\n",
    "    print(f\"Params: {params}; Nr of clusters: {len(pc.partition.sizes())}\")\n",
    "    print(\"#\" * 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore based on Decision Criteria\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cluster_coherence_analyzer(\n",
    "    analysis_data_dict,\n",
    "    print_n_random_titles,\n",
    "    word1,\n",
    "    word2,\n",
    "    print_word_string=True,\n",
    "    save_to_file=True,\n",
    "    file_name=\"output/tables/cluster-explorer/ParamsTest/cluster_analysis_output.txt\",\n",
    "    include_all_params_status=True,  # New argument to control the inclusion of all params status\n",
    "):\n",
    "    \"\"\"\n",
    "    Analyzes clusters for coherence based on the presence of specific words, prints details about matching clusters,\n",
    "    highlights if multiple hits are found within the same parameters, and optionally saves the output to a text file.\n",
    "\n",
    "    Args:\n",
    "    - analysis_data_dict (dict): Dictionary containing analysis data with keys as parameters (alpha, k, res) and values\n",
    "                                 as a list containing the summary DataFrame and a list of DataFrames for each cluster.\n",
    "    - print_n_random_titles (int): Number of random titles to print from matching clusters.\n",
    "    - word1 (str), word2 (str): Words to search for within the cluster's topic words.\n",
    "    - print_word_string (bool): Whether to print the concatenated string of topic words.\n",
    "    - save_to_file (bool): Whether to save the printed output to a text file.\n",
    "    - file_name (str): Name of the file to save the output to, if save_to_file is True.\n",
    "    \"\"\"\n",
    "    no_hit_params = []\n",
    "    output = []  # List to capture the output\n",
    "    output.append(f\"Words: {word1.upper()} {word2.upper()}\\n\")\n",
    "\n",
    "    for params, (summary, titles_df_list) in analysis_data_dict.items():\n",
    "        hits_per_param = 0  # Counter for hits within the same parameters\n",
    "        nr_of_clusters = len(titles_df_list)\n",
    "\n",
    "        for i, row in summary.iterrows():\n",
    "            Nr_of_Pubs = row[\"Nr of Pubs\"]\n",
    "            row_words_str = \" \".join([str(value) for value in row[4:]])\n",
    "            if word1 in row_words_str and word2 in row_words_str:\n",
    "                if hits_per_param == 0:\n",
    "                    output.append(\"=\" * 130)\n",
    "                hits_per_param += 1  # Increment the hit counter\n",
    "                output.append(\n",
    "                    f\"Found in {params};    Total Clusters: {nr_of_clusters};     Cluster: {row[0]};      Nr of Pubs in Cluster: {Nr_of_Pubs}\"\n",
    "                )\n",
    "                if print_word_string:\n",
    "                    output.append(row_words_str)\n",
    "                if print_n_random_titles != 0:\n",
    "                    output.append(\"RANDOM TITLES\")\n",
    "                    titles = titles_df_list[i][\"title_random\"].head(\n",
    "                        print_n_random_titles\n",
    "                    )\n",
    "                    output.extend([tit for tit in titles])\n",
    "                if hits_per_param > 0:\n",
    "                    output.append(\"- \" * 65)\n",
    "\n",
    "        if hits_per_param > 1:  # Check if there were multiple hits for the parameters\n",
    "            output.append(f\"** Multiple hits found in parameters: {params} **\\n\")\n",
    "        elif hits_per_param == 0 and include_all_params_status:\n",
    "            no_hit_params.append(params)\n",
    "\n",
    "    output.append(f\"No hits found for Parameters: \\n{set(no_hit_params)}\\n\")\n",
    "\n",
    "    # Convert list to string\n",
    "    output_str = \"\\n\".join(output)\n",
    "\n",
    "    # Print the output to console\n",
    "    print(output_str)\n",
    "\n",
    "    # Optionally save to file\n",
    "    if save_to_file:\n",
    "        with open(file_name, \"w\") as file:\n",
    "            file.write(output_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Decision Criteria:\n",
    "\n",
    "1. premature ejaculation and sexual dysfunction be distinct clusters\n",
    "2. Mixing of anxiety and panic disorder?\n",
    "3. Diabetic Mice VS Body Weight and Diabetes VS depression treatment in diabetics?\n",
    "4. Use alpha = 0.3 (look at alpha0.5_k10_res0.002); topics: 50, 133 -> very mixed words, little coherence.Not the case for alpha0.3_k10_res0.002 (here, sexual dys and premature ejac single topic though)\n",
    "5. Zimeldine (1 or two topics)\n",
    "6. Pregnancy related topics.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Words: EJAC DYSF\n",
      "\n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k15_res0.01;    Total Clusters: 379;     Cluster: 328;      Nr of Pubs in Cluster: 24\n",
      " sexual paroxetine male rats male rats irisin dysfunction sexual dysfunction induced ethyl acetate fraction ethyl acetate acetate fraction acetate ethyl fraction levels frequency latency ejaculatory paroxetine induced\n",
      "RANDOM TITLES\n",
      "Management of drug induced sexual dysfunction in male rats by ethyl acetate fraction of onion\n",
      "Irisin ameliorates male sexual dysfunction in paroxetine-treated male rats\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "No hits found for Parameters: \n",
      "{'alpha0.3_k15_res0.006', 'alpha0.5_k15_res0.006', 'alpha0.3_k10_res0.002', 'alpha0.5_k20_res0.01', 'alpha0.3_k20_res0.006', 'alpha0.3_k10_res0.006', 'alpha0.3_k20_res0.01', 'alpha0.5_k20_res0.006', 'alpha0.3_k20_res0.02'}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "analysis_data_dict = graph_summary_df_dict\n",
    "print_n_random_titles = 5\n",
    "word1 = \"ejac\"\n",
    "word2 = \"dysf\"\n",
    "print_word_string = True\n",
    "save_to_file = True\n",
    "file_name = \"../output/tables/cluster-explorer/ParamsTest/cluster_analysis_output.txt\"\n",
    "\n",
    "cluster_coherence_analyzer(\n",
    "    analysis_data_dict=analysis_data_dict,\n",
    "    print_n_random_titles=print_n_random_titles,\n",
    "    word1=word1,\n",
    "    word2=word2,\n",
    "    print_word_string=print_word_string,\n",
    "    save_to_file=save_to_file,\n",
    "    file_name=file_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Words: COVID \n",
      "\n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k10_res0.002;    Total Clusters: 155;     Cluster: 121;      Nr of Pubs in Cluster: 65\n",
      " covid patients serotonin fluvoxamine serotonin reuptake inhibitors inhibitors reuptake inhibitors fam lpla2 serotonin reuptake reuptake treatment grapefruit juice fluvoxamine juice fluvoxamine interaction risky juice fluvoxamine fluvoxamine interaction risky juice fluvoxamine interaction grapefruit\n",
      "RANDOM TITLES\n",
      "Re: ‘efficacy and safety of selective serotonin reuptake inhibitors in COVID-19 management: a systematic review and meta-analysis’ by Deng et al.\n",
      "Grapefruit juice-fluvoxamine interaction: Is it risky or not? [7]\n",
      "Can selective serotonin reuptake inhibitors have a neuroprotective effect during COVID-19?\n",
      "Neuropsychiatric Drugs Against COVID-19: What is the Clinical Evidence?\n",
      "A randomized multicenter clinical trial to evaluate the efficacy of melatonin in the prophylaxis of SARS-CoV-2 infection in high-risk contacts (MeCOVID Trial): A structured summary of a study protocol for a randomised controlled trial\n",
      "When Virtual Screening Yields Inactive Drugs: Dealing with False Theoretical Friends\n",
      "A fluorogenic phospholipid for the detection of lysosomal phospholipase A2 activity\n",
      "Drug repurposing of selective serotonin reuptake inhibitors: Could these drugs help fight COVID-19 and save lives?\n",
      "The role of human C5a as a non-genomic target in corticosteroid therapy for management of severe COVID19\n",
      "Pharmacological treatment of patients with mild to moderate COVID-19: A comprehensive review\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k10_res0.006;    Total Clusters: 362;     Cluster: 328;      Nr of Pubs in Cluster: 21\n",
      " treated ebov tolcapone covid reuptake patients psv inpatients serotonin clinical evaluation dependent clearance decrements clearance decrements fluvoxamine clearance time dependent clearance inpatients clinical time dependent time inpatients clinical evaluation\n",
      "RANDOM TITLES\n",
      "Time-dependent clearance decrements of fluvoxamine in depressed inpatients: A clinical evaluation [16]\n",
      "Selective Serotonin Reuptake Inhibitors may Influence COVID-19 Prognosis through Antioxidant and Cytoprotective Pathways Mediated by Sigma 1 Receptor Agonism\n",
      "Can selective serotonin reuptake inhibitors/serotonin and norepinephrine reuptake inhibitor antidepressants decrease the “cytokine storm” in the course of COVid-19 pneumonia?\n",
      "Tolcapone Potently Inhibits Seminal Amyloid Fibrils Formation and Blocks Entry of Ebola Pseudoviruses\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k15_res0.006;    Total Clusters: 241;     Cluster: 229;      Nr of Pubs in Cluster: 24\n",
      " covid fluoxetine early cvb4 treatment covid conditions fluvoxamine psychiatric medications fluvoxamine psychiatric conditions medications covid medications covid conditions psychiatric psychiatric medications psychiatric medications covid repurposing fluvoxamine psychiatric repurposing fluvoxamine medications repurposing fluvoxamine progressive encephalitis\n",
      "RANDOM TITLES\n",
      "Repurposing fluvoxamine, and other psychiatric medications, for COVID-19 and other conditions\n",
      "Persistent infection of human pancreatic cells with Coxsackievirus B4 is cured by fluoxetine\n",
      "Combined Treatment of Progressive Encephalitis in an X-linked Agammaglobulinemia Patient\n",
      "Efficacy and safety of selective serotonin reuptake inhibitors in COVID-19 management: author's response\n",
      "Hypoxia may be a determinative factor in COVID-19 progression\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k15_res0.01;    Total Clusters: 379;     Cluster: 346;      Nr of Pubs in Cluster: 18\n",
      " concerns covid covid rationale covid rationale concerns inhibitors inhibitors covid inhibitors covid rationale rationale rationale concerns repurposing repurposing selective repurposing selective serotonin reuptake reuptake inhibitors reuptake inhibitors covid selective selective serotonin selective serotonin reuptake serotonin serotonin reuptake\n",
      "RANDOM TITLES\n",
      "Repurposing Selective Serotonin Reuptake Inhibitors for COVID-19: Rationale and Concerns\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "Found in alpha0.3_k15_res0.01;    Total Clusters: 379;     Cluster: 353;      Nr of Pubs in Cluster: 15\n",
      " dispensing pandemic covid health medication dispensing mental health medication mental medications patterns covid pandemic dispensing patterns time series series analysis interrupted time series slope interrupted immediate interrupted time examined\n",
      "RANDOM TITLES\n",
      "Dispensing patterns of mental health medications before and during the COVID-19 pandemic in Alberta, Canada: An interrupted time series analysis\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "** Multiple hits found in parameters: alpha0.3_k15_res0.01 **\n",
      "\n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k20_res0.006;    Total Clusters: 190;     Cluster: 178;      Nr of Pubs in Cluster: 33\n",
      " covid fluvoxamine patients infection outpatients covid stand fluvoxamine outpatients covid fluvoxamine outpatients outpatients covid stand covid stand associated outpatients ceramide acute ssris dystonia associated dystonia metoclopramide acute dystonia fluvoxamine metoclopramide\n",
      "RANDOM TITLES\n",
      "Acute dystonia associated with fluvoxamine-metoclopramide [4]\n",
      "A randomized multicenter clinical trial to evaluate the efficacy of melatonin in the prophylaxis of SARS-CoV-2 infection in high-risk contacts (MeCOVID Trial): A structured summary of a study protocol for a randomised controlled trial\n",
      "Efficacy and safety of selective serotonin reuptake inhibitors in COVID-19 management: a systematic review and meta-analysis\n",
      "Ongoing use of SSRIs does not alter outcome in hospitalized COVID-19 patients: A retrospective analysis\n",
      "Fluvoxamine for outpatients with COVID-19: where do we stand?\n",
      "Beneficial effects of fluvoxamine for chorea in a patient with Huntington's disease: A case report\n",
      "Identification of Host-Targeted Small Molecules That Restrict Intracellular Mycobacterium tuberculosis Growth\n",
      "Use of physicochemical calculation of pKa and CLogP to predict phospholipidosis-inducing potential: A case study with structurally related piperazines\n",
      "Outcomes and Laboratory and Clinical Findings of Asthma and Allergic Patients Admitted With Covid-19 in a Spanish University Hospital\n",
      "Acid sphingomyelinase inhibitors normalize pulmonary ceramide and inflammation in cystic fibrosis\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "==================================================================================================================================\n",
      "Found in alpha0.3_k20_res0.02;    Total Clusters: 563;     Cluster: 104;      Nr of Pubs in Cluster: 117\n",
      " panic disorder treatment panic disorder sertraline agoraphobia study scbt primary care care disorder agoraphobia panic disorder agoraphobia risk patients primary randomized pd placebo covid respiratory\n",
      "RANDOM TITLES\n",
      "Erratum: A randomized, naturalistic, parallel-group study for the long-term treatment of panic disorder with clonazepam or paroxetine (Journal of Clinical Psychopharmacology (2012) 32 (120-126))\n",
      "Panic and pandemic: Review of the literature on the links between panic disorder and the SARS-CoV-2 epidemic\n",
      "Psychiatric treatment received by primary care patients with panic disorder with and without agoraphobia\n",
      "A randomized trial of sertraline, self-administered cognitive behavior therapy, and their combination for panic disorder\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "Found in alpha0.3_k20_res0.02;    Total Clusters: 563;     Cluster: 359;      Nr of Pubs in Cluster: 37\n",
      " tryptophan patients ht coronavirus anosmia ageusia deficiency anosmia ageusia coronavirus patients covid depletion receptor severity serotonin chemesthesis ht deficiency disease chemesthetic increased precursor\n",
      "RANDOM TITLES\n",
      "Does serotonin deficiency lead to anosmia, ageusia, dysfunctional chemesthesis and increased severity of illness in COVID-19?\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "Found in alpha0.3_k20_res0.02;    Total Clusters: 563;     Cluster: 476;      Nr of Pubs in Cluster: 16\n",
      " sars sars cov cov srt psv infection sars cov infection infection sars cov psv coronavirus effective protein sars cov psv treatment exhibits variants inhibits cell covid activity\n",
      "RANDOM TITLES\n",
      "Sertraline Is an Effective SARS-CoV-2 Entry Inhibitor Targeting the Spike Protein\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "Found in alpha0.3_k20_res0.02;    Total Clusters: 563;     Cluster: 480;      Nr of Pubs in Cluster: 15\n",
      " women ci men women ci men ci years mean age years adults adults mean adults mean age age age years age years female years female mean january women ci men prescriptions benzodiazepine covid\n",
      "RANDOM TITLES\n",
      "Trends in the Use of Benzodiazepines, Z-Hypnotics, and Serotonergic Drugs among US Women and Men before and during the COVID-19 Pandemic\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "** Multiple hits found in parameters: alpha0.3_k20_res0.02 **\n",
      "\n",
      "==================================================================================================================================\n",
      "Found in alpha0.5_k15_res0.006;    Total Clusters: 298;     Cluster: 274;      Nr of Pubs in Cluster: 17\n",
      " covid fluvoxamine fluvoxamine treatment fluvoxamine treatment covid treatment treatment covid None None None None None None None None None None None None None None\n",
      "RANDOM TITLES\n",
      "Fluvoxamine for the treatment of COVID-19\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "Found in alpha0.5_k15_res0.006;    Total Clusters: 298;     Cluster: 296;      Nr of Pubs in Cluster: 1\n",
      " dsps adolescents increase dsps adolescents covid pandemic covid pandemic number number dsps paracetamol self years year age study substance strong increase strong second poisonings\n",
      "RANDOM TITLES\n",
      "Significant Increase in Deliberate Self-Poisonings Among Adolescents During the Second Year of the COVID-19 Pandemic\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "** Multiple hits found in parameters: alpha0.5_k15_res0.006 **\n",
      "\n",
      "==================================================================================================================================\n",
      "Found in alpha0.5_k20_res0.006;    Total Clusters: 231;     Cluster: 119;      Nr of Pubs in Cluster: 103\n",
      " sodium hydrochloride clinical trials drug alfa ebv hiv clinical trials viral insulin living covid gateways clinical gateways gateways clinical trials viral load living hiv reactivation showed\n",
      "RANDOM TITLES\n",
      "Gateways to clinical trials: December 2004\n",
      "Reactivation of the epstein-Barr virus leading to acute liver failure in a patient living with HIV\n",
      "Drug repurposing for COVID-19: current evidence from randomized controlled adaptive platform trials and living systematic reviews\n",
      "Gateways to clinical trials.\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "==================================================================================================================================\n",
      "Found in alpha0.5_k20_res0.01;    Total Clusters: 375;     Cluster: 324;      Nr of Pubs in Cluster: 13\n",
      " covid mortality patients ssris hospital ci exposed retrospective use hospital mortality study ssri selective serotonin cohort study effect reuptake inhibitors disease selective serotonin selective serotonin reuptake\n",
      "RANDOM TITLES\n",
      "Association between Selective Serotonin Reuptake Inhibitors Prevalent Use and COVID-19-Related Mortality: A Retrospective Cohort Study\n",
      "- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - \n",
      "No hits found for Parameters: \n",
      "{'alpha0.3_k20_res0.01'}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "analysis_data_dict = graph_summary_df_dict\n",
    "print_n_random_titles = 10\n",
    "word1 = \"covid\"\n",
    "word2 = \"\"\n",
    "print_word_string = True\n",
    "save_to_file = True\n",
    "file_name = \"../output/tables/cluster-explorer/ParamsTest/cluster_analysis_output.txt\"\n",
    "\n",
    "cluster_coherence_analyzer(\n",
    "    analysis_data_dict=analysis_data_dict,\n",
    "    print_n_random_titles=print_n_random_titles,\n",
    "    word1=word1,\n",
    "    word2=word2,\n",
    "    print_word_string=print_word_string,\n",
    "    save_to_file=save_to_file,\n",
    "    file_name=file_name,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SUICIDE\n",
    "params_scoring_suicide = {\n",
    "    \"alpha0.3_k10_res0.002\": \"1\",\n",
    "    \"alpha0.3_k10_res0.006\": \"0 (too many suicide clusters, very inconsistent)\",\n",
    "    \"alpha0.3_k15_res0.006\": \"1 (very clear singel suicide cluster)\",\n",
    "    \"alpha0.3_k15_res0.01\": \"0 (conflates overdoses and suicide; too many clusters on it; 1 clear)\",\n",
    "    \"alpha0.3_k20_res0.006\": \"1 (one good suicide, one unrelated)\",\n",
    "    \"alpha0.3_k20_res0.01\": \"1 (two clear suicide clusters; should be same)\",\n",
    "    \"alpha0.3_k20_res0.02\": \"0 (akathisia cluster; boderline cluster; no clear bigger suicide cluster)\",\n",
    "    \"alpha0.5_k15_res0.006\": \"0 no good suicide cluster\",\n",
    "    \"alpha0.5_k20_res0.006\": \"1 (one clear suicide cluster, one overdose/drug)\",\n",
    "    \"alpha0.5_k20_res0.01\": \"0 first biggest cluster is very bad\",\n",
    "}\n",
    "# pregnancy\n",
    "params_scoring_pregnancy = {\n",
    "    \"alpha0.3_k10_res0.002\": \"1 (single cluster, coherent)\",\n",
    "    \"alpha0.3_k10_res0.006\": \"0 (pharma pregnancy, narcolepsy, lectation, neonatal, TOO MANY)\",\n",
    "    \"alpha0.3_k15_res0.006\": \"0 (4 clusters, not one clear preggo cluster)\",\n",
    "    \"alpha0.3_k15_res0.01\": \"0 (1 big trash topic; pretty bad)\",\n",
    "    \"alpha0.3_k20_res0.006\": \"0 (one small preggo cluster)\",\n",
    "    \"alpha0.3_k20_res0.01\": \"1 (bit too split up, but sensical, [pretty good])\",\n",
    "    \"alpha0.3_k20_res0.02\": \"0 (creation seems wieird.)\",\n",
    "    \"alpha0.5_k15_res0.006\": \"0 (biggest cluster sucks. creation went weird)\",\n",
    "    \"alpha0.5_k20_res0.006\": \"0 (first biggest cluster is weird)\",\n",
    "    \"alpha0.5_k20_res0.01\": \"0 (small cluster, seem sensical tho)\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Params: alpha0.3_k10_res0.002; Nr of clusters: 154\n",
      "##################################################\n",
      "Params: alpha0.3_k20_res0.01; Nr of clusters: 299\n",
      "##################################################\n",
      "Params: alpha0.3_k20_res0.006; Nr of clusters: 190\n",
      "##################################################\n"
     ]
    }
   ],
   "source": [
    "iterations = 50\n",
    "quality_values = []\n",
    "nr_clusters = []\n",
    "partitions = []\n",
    "graph_summary_df_dict = {}\n",
    "\n",
    "last_selection_params = [\n",
    "    \"alpha0.3_k10_res0.002\",\n",
    "    \"alpha0.3_k20_res0.01\",\n",
    "    \"alpha0.3_k20_res0.006\",\n",
    "]\n",
    "\n",
    "\n",
    "for params in last_selection_params:\n",
    "    G = params_graph_dict[params.split(\"_res\")[0]]\n",
    "    resolution = pd.to_numeric(params.split(\"res\")[1])\n",
    "    column_name = f\"cluster_{params}\"\n",
    "    pc = PartitionCreator(G, df)\n",
    "    pc.create_partition_from_cmpvertexpartition(\n",
    "        n_iterations=iterations,\n",
    "        resolution_parameter=resolution,\n",
    "        verbose=False,\n",
    "        cluster_colunm_name=column_name,\n",
    "        centrality_column_name=f\"centrality_{params}\",\n",
    "    )\n",
    "    #############################\n",
    "    #############################\n",
    "    ce = CommunityExplorer(pc.df, f\"cluster_{params}\")\n",
    "    ce.create_cluster_sheets(sort=True, n=25, sort_column=f\"centrality_{params}\")\n",
    "    sorted_sheets = [s.reset_index(drop=True) for s in ce.sheets]\n",
    "    # suffix\n",
    "    sorted_sheets = [s.add_suffix(f\"_sorted\") for s in sorted_sheets]\n",
    "    ce.create_cluster_sheets(sort=False, n=25, sort_column=f\"centrality_{params}\")\n",
    "    unsorted_sheets = [s.reset_index(drop=True) for s in ce.sheets]\n",
    "    # suffix\n",
    "    unsorted_sheets = [s.add_suffix(f\"_random\") for s in unsorted_sheets]\n",
    "    comb_sheets = [\n",
    "        pd.concat([sorted_sheets[i], unsorted_sheets[i]], axis=1)\n",
    "        for i in range(len(sorted_sheets))\n",
    "    ]\n",
    "    ce.create_summary_sheet()\n",
    "    df_summary = ce.df_summary\n",
    "    graph_summary_df_dict[f\"{params}\"] = [df_summary, comb_sheets]\n",
    "    # save to excel\n",
    "    filepath = (\n",
    "        f\"../output/tables/cluster-explorer/FinalSelect/SingleSolExplorer_{params}.xlsx\"\n",
    "    )\n",
    "    save_to_excel_with_hyperlinks(filepath, df_summary, comb_sheets)\n",
    "    print(f\"Params: {params}; Nr of clusters: {len(pc.partition.sizes())}\")\n",
    "    print(\"#\" * 50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FINAL SELECTION\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## alpha0.3_k10_res0.002; Nr of clusters: 154\n",
    "\n",
    "1. \"alpha0.3_k10_res0.002\": \"1 - suicide\",\n",
    "2. \"alpha0.3_k10_res0.002\": \"1 (single cluster, coherent) - pregnancy\",\n",
    "\n",
    "##### EVAL TOP 5 CLSUTERS\n",
    "\n",
    "0. Very clear memory clsuter; pharma and not; 1199 pubs\n",
    "1. Receptor binding; brain; guess its good\n",
    "2. QTC prolongation; cardiovascular; good\n",
    "3. astrocytes; pharma; good\n",
    "4. Binding; platelets check difference to 2; (kinda good)\n",
    "5. Pulmonary hypertension; bluthochdruck; (kinda good)\n",
    "\n",
    "## alpha0.3_k20_res0.01; Nr of clusters: 299\n",
    "\n",
    "1. \"alpha0.3_k20_res0.01\": \"1 (two clear suicide clusters; should be same) - suicide\",\n",
    "2. \"alpha0.3_k20_res0.01\": \"1 (bit too split up, but sensical, [pretty good]) - pregnancy\",\n",
    "3. Good elderly cluster\n",
    "4. Good Fall cluster (in elderly)\n",
    "5. Good single seasonal cluster\n",
    "6. good withdrawal/discontinuation cluster\n",
    "\n",
    "##### EVAL TOP 5 CLSUTERS\n",
    "\n",
    "0. Learning; Memory;hippocampus; fear (avoidance learning); Ok Good\n",
    "1. Prolactin; Pharmacology; ok Good\n",
    "2. Cardiac,arrythmia, qt prolongation (very good)\n",
    "3. pregnancy; placenta, a preggo pharma cluster\n",
    "4. SSRI mechanisms; Pharmacology; Depression treatment\n",
    "5. Pharmacology; 5-HT; receptor; SERT(serotonintransporter),\n",
    "\n",
    "## alpha0.3_k20_res0.006; Nr of clusters: 190\n",
    "\n",
    "1. \"alpha0.3_k20_res0.01\": \"1 (two clear suicide clusters; should be same)- suicide\",\n",
    "2. \"alpha0.3_k20_res0.01\": \"1 (bit too split up, but sensical, [pretty good])- pregnancy\",\n",
    "\n",
    "##### EVAL TOP 5 CLSUTERS\n",
    "\n",
    "0. Memory; Cognition; Learning;\n",
    "1. Receptor; Animal studies; IN vivo; autoreceptor;\n",
    "2. cardiac; qt prolongation\n",
    "3. astrocytes; pharmaco\n",
    "4. pharmacological ssri properties; 5 ht; receptors\n",
    "5. vascular; rat; unclear;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_given_labels(df_dict, params, words_to_remove):\n",
    "    cols = [col for col in df_dict[params][0].columns if col.startswith(\"Word_\")]\n",
    "    df_dict[params][0][\"Given Label\"] = df_dict[params][0][cols].apply(\n",
    "        lambda row: list(set(row.values)), axis=1\n",
    "    )\n",
    "    givenlabel = []\n",
    "    for label in df_dict[params][0][\"Given Label\"]:\n",
    "        label = [word for word in label[:10] if word not in words_to_remove]\n",
    "        givenlabel.append(label)\n",
    "        print(label)\n",
    "    df_dict[params][0][\"Given Label\"] = givenlabel\n",
    "\n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['mg', 'amitriptyline', 'performance', 'healthy', 'learning', 'patients', 'sleep', 'ht']\n",
      "['opioid', 'antidepressant', '5ht', 'brain', 'ne', 'pindolol', 'serotonin', 'ht 1a', 'ht']\n",
      "['qt prolongation', 'mg', 'prolongation', 'cardiac', 'interval', 'qtc interval', 'antidepressants']\n",
      "['glial', 'glioma', 'protein', 'antidepressant', '5ht', 'brain', 'serotonin', 'ht', 'activity']\n",
      "['patients', 'dat', 'brain', 'serotonin', 'controls', 'ht', 'transporter']\n",
      "['tachycardia', 'pulmonary', 'induced', 'ph', 'pulmonary arterial', 'epididymis', 'sperm', 'rats', 'serotonin']\n",
      "['amitriptyline', 'blind', 'depressive', 'patients', 'major depressive', 'major', 'study']\n",
      "['release', 'fenfluramine', 'behavioral', 'amphetamine', 'raphe', 'rats', '5ht', 'brain', 'serotonin']\n",
      "['depressed', 'patients', 'efficacy', 'reuptake inhibitors', 'norepinephrine', 'serotonin', 'ht', 'inhibitor', 'pharmacological']\n",
      "['magnesium', 'eeg', 'model', 'patients', 'mgkg', 'antidepressant', 'rats', 'ht']\n",
      "['prolactin', 'fenfluramine', 'induced', 'serotonin', 'ht', 'secretion', 'galactorrhoea']\n",
      "['movement', 'bruxism', 'sertraline bruxism', 'induced', 'dyskinesia', 'disorder fluoxetine', 'patient', 'serotonin']\n",
      "['ci', 'antidepressant', 'serotonin', 'expression', 'antidepressants', 'increased', 'tumor', 'apoptosis']\n",
      "['paroxetine personality', 'effect paroxetine personality', 'personality', 'patients', 'disorder', 'bpd', 'ht']\n",
      "['cocaine', 'behavioral', 'induced', 'methylphenidate', 'htp', 'mgkg', 'dom', 'ht']\n",
      "['ibs', 'release', 'μm', 'induced', 'reflux', 'patients', 'serotonin']\n",
      "['3h', 'hz', 'release', 'tritium', '5ht', 'brain', 'serotonin', 'ht', 'efflux', 'synaptosomes']\n",
      "['fear', 'anxiolytic', 'ht1a', 'mgkg', 'anxiolytic like', 'rats', 'ht', 'anxiety']\n",
      "['placental', 'infants', 'studies', 'outcomes', 'pregnancy', 'serotonin', 'maternal', 'ssri', 'exposure', 'women']\n",
      "['pain', 'female', 'rats', 'brain', 'serotonin', 'maternal', 'ssri', 'ht']\n",
      "['use paroxetine', 'induced', 'given patients', 'patients', 'efficacy', 'antidepressant', 'given', 'disorder']\n",
      "['perspective', 'capillary', 'studies', 'patients', 'chiral', 'antidepressant', 'risks citalopram', 'ndcita']\n",
      "['induced', 'mgkg', 'antidepressant', 'hypothermia', 'rats', 'ht', 'repeated', 'receptor', 'α1', 'dopamine']\n",
      "['induced', 'antidepressant', '5ht', 'serotonin', 'compounds', 'ht', 'antidepressants', 'transporter', 'sert']\n",
      "['pain', 'antinociceptive', 'mgkg', 'serotonin', 'activity', 'antidepressants', 'chronic']\n",
      "['μm', 'potential', 'sensitive', 'sleep', 'hermaphrodita', '5ht', 'motoneurons', 'serotonin', 'wa']\n",
      "['release', 'postnatal', 'na', 'brain', 'serotonin', 'ht', 'transporter', 'sert', 'dopamine', 'da']\n",
      "['type', 'patients', 'significant', 'insulin', 'glucose', 'study', 'diabetes mellitus', 'sertraline']\n",
      "['axis', 'acth', 'mg', 'stress', 'patients', 'antidepressant', 'hpa', 'oral']\n",
      "['food intake', 'sibutramine', 'food', 'obese', 'intake', 'patients', 'antidepressant', 'rats', 'therapy']\n",
      "['human plasma', 'liquid', 'phase', 'detection', 'performance', 'high', 'electrochemical', 'norfluoxetine']\n",
      "['opioid', 'olfactory', 'mgkg', 'antidepressant', 'rats', 'serotonin', 'activity', 'bh4']\n",
      "['drug interactions', 'interactions', 'μm', 'palbociclib', 'cytochrome', 'warfarin', 'pharmacokinetic']\n",
      "['hepatitis', 'induced', 'patients', 'fluvoxamine liver enzymes']\n",
      "['syndrome', 'rem', 'patients', 'sleep', 'antidepressant', 'disorder', 'restless', 'time']\n",
      "['induced seizures', 'induced', 'methylphenidate', 'patients', 'mgkg', 'brain', 'serotonin']\n",
      "['locomotor', 'removal', 'ph', 'fish', 'uv', 'exposure', 'biosolids', 'test', 'like']\n",
      "['mdd', 'trazodone', 'depressive', 'studies', 'patients', 'efficacy', 'antidepressant', 'disorder', 'serotonin']\n",
      "['depressed', 'methylphenidate', 'depressive', 'patients', 'selective serotonin reuptake', 'antidepressant', 'serotonin', 'older']\n",
      "['mg', 'recurrent', 'long', 'venlafaxine', 'patients', 'long term', 'efficacy']\n",
      "['social anxiety disorder', 'social anxiety', 'phobia', 'sad', 'generalized', 'patients', 'therapy', 'disorder']\n",
      "['thyroid function', 'depressive', 'patients', 'antidepressant', 'function', 'disorder', 'depressive disorder', 'major depressive']\n",
      "['serotonin syndrome', 'syndrome', 'patients', 'symptoms', 'patient', 'case report', 'serotonin']\n",
      "['compulsive', 'obsessive compulsive', 'emdr', 'patients', 'obsessive', 'therapy', 'disorder', 'compulsive disorder']\n",
      "['fst', 'differences', 'mgkg', 'immobility', 'antidepressant', 'swimming', 'rats', 'serotonin']\n",
      "['pain', 'mouth', 'syndrome', 'patients', 'oral', 'therapy', 'burning', 'mouth syndrome', 'antidepressants']\n",
      "['dependence', 'acamprosate', 'studies', 'intake', 'patients', 'ethanol', 'rats', 'alcohol']\n",
      "['asymmetric', 'process', 'reaction', 'hydrochloride', 'enantioselective', 'chiral', 'antidepressant']\n",
      "['ssri induced', 'induced mania', 'induced', 'adolescent', 'hypomania', 'mania induced fluoxetine']\n",
      "['ssri withdrawal', 'withdrawal', 'ssri response', 'letters', 'news', 'ssri', 'stopped', 'ads questioned', 'ssri ads questioned', 'review stopped']\n",
      "['circadian', 'depressive', 'sad', 'agomelatine', 'patients', 'light', 'affective disorder', 'antidepressant']\n",
      "['safety', 'fluoxetine desipramine', 'fluoxetine overdose', 'patients', 'efficacy', 'safety efficacy', 'sleep', 'overdose']\n",
      "['effort', 'rate', 'stress', 'reinforcement', 'mgkg', 'antidepressant', 'rats', 'reversal', 'serotonin', 'ht']\n",
      "['effective', 'depressive', 'patients', 'psychiatrists', 'antidepressant', 'therapy', 'ssri', 'mirtazapine', 'antidepressants']\n",
      "['galantamine', 'evidence', 'patients', 'dementia', 'symptoms', 'agitation', 'ssri', 'antidepressants']\n",
      "['gal', 'galanin', 'stress', 'rem', 'sleep', 'antidepressant', 'brain']\n",
      "['reactions', 'amitriptyline', 'induced', 'monoamine oxidase', 'patients', 'antidepressant', 'serotonin', 'antidepressants', 'azoles']\n",
      "['bupropion', 'female', 'induced', 'patients', 'dysfunction', 'fluoxetine induced sexual', 'rats', 'fluoxetine induced']\n",
      "['new', 'type', 'cataplexy', 'patients', 'onset', 'sleep', 'pregnancy', 'sodium', 'sodium oxybate', 'solriamfetol']\n",
      "['pd', 'panic', 'effective', 'patients', 'benzodiazepines', 'therapy', 'disorder']\n",
      "['channels', 'release', 'μm', 'herg', 'voltage', 'neuronal', 'currents', 'norfluoxetine']\n",
      "['depressed', 'depressive', 'patients', 'sleep', 'tryptophan', 'serotonin', 'mood', 'atd']\n",
      "['activin', 'stress', 'ltp', 'phosphorylation', 'neurogenesis', 'antidepressant', 'rats', 'expression']\n",
      "['treatment depression', 'pd', 'treatment depression parkinsons', 'depression parkinsons disease', 'patients', 'symptoms', 'patient', 'parkinsons']\n",
      "['ham', 'mg', 'patients', 'major depression', 'efficacy', 'symptoms', 'anxiety', 'major']\n",
      "['depressive', 'symptom', 'patients', 'antidepressant', 'symptoms', 'therapy', 'remission', 'cognitive therapy']\n",
      "['stroke', 'psd', 'ischaemic', 'patients', 'therapy', 'motor recovery', 'recovery']\n",
      "['related', 'events', 'safety', 'data', 'patients', 'database', 'medicines', 'using', 'medication', 'drug']\n",
      "['rich', 'thioridazine', 'lysosomal', 'ratios', 'antidepressants', 'trapping', 'drug', 'tissues', 'drugs']\n",
      "['severe', 'induced', 'patients', 'hyponatremia', 'hyponatraemia', 'patient', 'sodium', 'ssri', 'elderly']\n",
      "['dog', 'phase', 'hsha', 'bias', 'cent', 'prescribed', 'study', 'mice', 'dogs']\n",
      "['acute angle closure', 'serotonin reuptake', 'acute angle', 'angle closure', 'acute', 'selective serotonin reuptake', 'closure', 'serotonin', 'selective serotonin']\n",
      "['anorexia', 'bulimia', 'patients', 'nervosa', 'patient', 'disorder', 'anorexia nervosa', 'serotonin', 'nervosa patient fluoxetine']\n",
      "['ssri induced', 'induced', 'studies', 'patients', 'disorder', 'time', 'serotonin', 'ssri', 'children']\n",
      "['n1', 'dependence', 'patients', 'ldaep', 'intensity', 'serotonin', 'auditory evoked', 'auditory', 'p50']\n",
      "['adsorption', 'lipid', 'fluoxetine hydrochloride', 'hydrochloride', 'solubility', 'salt', 'tf', 'using', 'fxt tf']\n",
      "['mdd', 'acute', 'patients', 'antidepressant', 'brain', 'serotonin', 'bias', 'connectivity', 'antidepressants']\n",
      "['mdd', 'ca1', 'ketamine', 'antidepressant', 'brain', 'pfc', 'ht', 'expression', 'alpha2']\n",
      "['cocaine', 'amperozide', 'dat', 'antidepressant', 'serotonin', 'derivatives', 'compounds', 'activity', 'test', 'receptor']\n",
      "['serotonin reuptake inhibitors', 'pharmacotherapy', 'serotonin reuptake', 'bdd', 'severity', 'patients', 'selective serotonin reuptake', 'delusional', 'therapy', 'disorder']\n",
      "['nlrp3 inflammasome', 'stress', 'nitrosylation', 'induced', 'nlrp3', 'rats', 'exd', 'pathway', 'expression']\n",
      "['activation', 'lymphocyte', 'stress', 'il', 'induced', 'patients', 'antidepressants', 'nsaids']\n",
      "['amitriptyline', 'effective', 'patients', 'headache', 'serotonin', 'inhibitor', 'chronic', 'migraine sertraline', 'drugs']\n",
      "['monotherapy', 'patients', 'antidepressant', 'therapy', 'disorder', 'ii', 'ofc', 'rapid']\n",
      "['recommendation sertraline', 'sertraline lustral', 'death sertraline', 'sertraline interactions', 'death', 'impending death', 'sertraline increased', 'lustral', 'cost sertraline', 'associated rhabdomyolysis']\n",
      "['stress', 'stress urinary', 'urinary retention', 'bladder', 'induced', 'behavioural', 'retention high', 'patient']\n",
      "['stress', 'induced', 'patients', 'mgkg', 'rats']\n",
      "['hivaids', 'treatment depression', 'patients hiv', 'patients hiv infection', 'patients', 'hiv', 'depression patients', 'trial']\n",
      "['data', 'self harm', 'suicidal', 'patients', 'antidepressant', 'harm', 'suicidal ideation', 'suicidality']\n",
      "['adhd', 'syndrome', 'patients', 'history', 'symptoms', 'disorder', 'ssri', 'apathy', 'anxiety']\n",
      "['ol', 'cd', 'controlled', 'negative symptoms schizophrenia', 'studies', 'negative symptoms', 'patients', 'symptoms', 'schizophrenia']\n",
      "['school refusal', 'refusal', 'label', 'school', 'dimensions', 'patients', 'antidepressant', 'anxiety']\n",
      "['rice', 'torsades pointes', 'stress', 'torsades', 'clinical trials', 'hydrochloride', 'signal', 'sodium', 'kno3', 'salt']\n",
      "['withdrawal', 'premenstrual', 'pms', 'syndrome', 'model', 'ethanol', 'symptoms', 'oral']\n",
      "['spectrum disorders', 'autism spectrum', 'spectrum', 'self injury', 'syndrome', 'repetitive', 'injury', 'asd', 'disorder']\n",
      "['autonomic', 'cardiovascular', 'antidepressant use', 'patients', 'antidepressant', 'symptoms', 'cardiac', 'patient', 'mortality']\n",
      "['compulsive', 'pathological gambling', 'buying', 'gamblers', 'patients', 'disorder', 'ssri', 'trial', 'pharmacological']\n",
      "['pathological crying', 'tbi', 'patients', 'injury', 'traumatic', 'brain', 'hpi', 'pharmacological']\n",
      "['bupropion', 'quit', 'quit date', 'smoking', 'antidepressants', 'drugs', 'date', 'cessation']\n",
      "['hallucinations', 'perception', 'perception disorder', 'persisting', 'venlafaxine', 'persisting perception disorder', 'symptoms', 'disorder', 'hallucinogen persisting perception']\n",
      "['pharmacogenetic', 'pgx', 'polymorphism', 'patients', 'brain', 'dosing', 'gene', 'testing', 'fold']\n",
      "['psychotropic drugs', 'medicine', 'patients', 'antidepressants', 'cannabis', 'psychotropic medications', 'psychotropic', 'drugs', 'ssris']\n",
      "['hair', 'induced', 'skin picking', 'patients', 'picking', 'fluvoxamine photosensitivity', 'disorder', 'drug']\n",
      "['augmentation fluoxetine treatment', 'hair', 'treatment trichotillomania', 'adolescent', 'olanzapine augmentation', 'ttm', 'disorder', 'augmentation fluoxetine', 'fluoxetine treatment trichotillomania']\n",
      "['pharmacotherapy', 'stress', 'tes', 'patients', 'traumatic', 'disorder', 'pharmacological', 'emotion', 'ssris']\n",
      "['data', 'syndrome', 'depressive', 'depressive symptoms', 'patients', 'symptoms', 'quality', 'anxiety']\n",
      "['dysthymia', 'minor depression', 'mean', 'patients', 'symptoms', 'disorder', 'pst pc']\n",
      "['market', 'european countries', 'generic', 'inhibitors', 'drug', 'drugs', 'trust', 'countries']\n",
      "['mg', 'polymyxin', 'gut', 'fluconazole', 'tabl', 'activity', 'biofilms', 'mg tabl']\n",
      "['costs', 'bupropion', 'cost effectiveness', 'patients', 'effectiveness', 'antidepressant', 'therapy']\n",
      "['mdd', 'symptom', 'antidepressant', 'symptoms', 'therapy', 'test', 'worsening', 'symptom worsening']\n",
      "['aneurysm', 'heroin', 'samples', 'hair', 'deaths', 'school', 'post', 'assay', 'drug', 'fatal']\n",
      "['depersonalization', 'fluoxetine depersonalization disorder', 'depersonalisation', 'depersonalisation disorder', 'syndrome', 'induced', 'patients', 'lamotrigine', 'syndrome induced fluoxetine']\n",
      "['mdd', 'rest interval', 'performance', 'depressive', 'studies', 'patients', 'symptoms', 'interval', 'ssri']\n",
      "['serotonin reuptake inhibitors', 'pain', 'serotonin reuptake', 'reuptake inhibitors fibromyalgia', 'patients', 'selective serotonin reuptake', 'reuptake inhibitors', 'therapy fibromyalgia', 'therapy']\n",
      "['autonomic', 'studies', 'ss', 'vasovagal syncope', 'patients', 'symptoms', 'patient', 'syncope']\n",
      "['fatigue', 'social support', 'agreement', 'patients', 'ms', 'phenomenon', 'multiple sclerosis', 'prescribed', 'drug']\n",
      "['patients', 'antidepressant', 'reuptake inhibitors', 'serotonin', 'mirtazapine', 'antidepressants', 'inhibitors', 'study', 'disorders']\n",
      "['withdrawal', 'benzodiazepine', 'discontinuation syndrome', 'events', 'syndrome', 'patients', 'symptoms', 'serotonin', 'ssri']\n",
      "['sertraline used ckd', 'ckd patients', 'used ckd', 'depressive', 'patients depression', 'patients', 'antidepressant', 'symptoms', 'hemodialysis']\n",
      "['treatment selective', 'fluoxetine elective', 'mutism', 'disorder', 'selective mutism', 'phenelzine', 'catatonia', 'fluoxetine elective mutism']\n",
      "['serotonin reuptake inhibitors', 'risky', 'serotonin reuptake', 'fam', 'fvx', 'patients', 'reuptake inhibitors', 'covid', 'serotonin']\n",
      "['mg', 'caused', 'expired escitalopram', 'premature ejaculation', 'pe', 'patients', 'dysfunction', 'ssri', 'nocturia caused', 'sexual']\n",
      "['evidence', 'states', 'patients', 'efficacy', 'reuptake inhibitors', 'symptoms', 'disorder', 'treatments']\n",
      "['mdd', 'mg', 'enantiomer', 'venlafaxine', 'patients', 'efficacy', 'results']\n",
      "['serotonin reuptake inhibitors', 'serotonin reuptake', 'fractures', 'reuptake inhibitors associated', 'bmd', 'reuptake inhibitors', 'serotonin', 'selective serotonin', 'ssri', 'bone']\n",
      "['breathlessness', 'pulmonary', 'patients', 'serotonin', 'anxiety', 'chronic', 'disease', 'disorders']\n",
      "['serotonin reuptake', 'male', 'studies', 'selective serotonin reuptake', 'mgkg', 'rats', 'sperm', 'selective serotonin']\n",
      "['treatment response', 'polymorphism', 'patients', 'antidepressant', 'genetic', 'serotonin', 'htr1a', 'polymorphisms']\n",
      "['severe', 'dizziness', 'studies', 'patients', 'serotonin', 'sertraline severe', 'sertraline severe tinnitus', 'anxiety']\n",
      "['hd', 'mouse', 'model', 'symptoms', 'patient', 'brain', 'therapist', 'sert', 'disease']\n",
      "['adsorption', 'water', 'showed', 'charged', 'smx', 'removal', 'mgg', 'kinetic']\n",
      "['ham', 'estrogen', 'ovx', 'depressive', 'differences', 'patients', 'mgkg', 'antidepressant']\n",
      "['mdd', 'methylfolate', 'vitd', 'depressive', 'patients', 'antidepressant', 'therapy', 'supplementation']\n",
      "['tamoxifen', 'breast cancer', 'patients', 'sleep', 'symptoms', 'therapy', 'menopausal', 'activity', 'hot']\n",
      "['sertraline risperidone', 'priapism', 'induced', 'priapism associated sertraline', 'priapism sertraline risperidone', 'patient', 'priapism associated', 'priapism sertraline']\n",
      "['hypericum', 'st johns', 'johns', 'st johns wort', 'wort', 'johns wort extract', 'efficacy', 'st', 'perforatum']\n",
      "['mdd', 'patients', 'antidepressant', 'disorder', 'expression', 'gene', 'tcreb', 'response', 'baseline']\n",
      "['channel congenital myasthenic', 'syndrome', 'patients', 'variants', 'nerve', 'scs', 'myasthenic', 'mutation', 'channel congenital']\n",
      "['transcranial', 'current stimulation', 'rtms', 'direct current', 'therapy', 'stimulation', 'neurofeedback', 'direct current stimulation']\n",
      "['cp', 'activation', 'induced', 'igf1r', 'megalin', 'ra', 'ahr', 'expression', 'receptor']\n",
      "['compulsive', 'obsessive compulsive', 'induced', 'disorders children', 'children adolescents', 'anxiety disorders children', 'maleate', 'obsessive']\n",
      "['cs', 'induced', 'mgkg', 'antidepressant', 'rats', 'amy', 'catalpol', 'activity', 'test', 'like']\n",
      "['sibutramine', 'food', 'samples', 'detection', 'dietary supplements', 'dietary', 'pills', 'drug', 'drugs', 'skin']\n",
      "['gateways clinical', 'clinical trials', 'hydrochloride', 'integrity', 'gateways clinical trials', 'calcium', 'prous', 'sodium', 'insulin', 'alfa']\n",
      "['syndrome', 'cerebral', 'induced', 'reversible cerebral', 'rcvs', 'vasoconstriction syndrome', 'carotidynia', 'cerebral vasoconstriction syndrome', 'fluoxetine induced']\n",
      "['ppis', 'ci', 'i2', 'colitis', 'patients', 'microscopic colitis', 'microscopic', 'control', 'ci i2', 'nsaids']\n",
      "['serotonin reuptake inhibitors', 'new', 'serotonin reuptake', 'new ht', 'new ht reuptake', 'selective serotonin reuptake', 'reuptake inhibitors', 'inhibitor better', 'selective serotonin', 'serotonin']\n",
      "['double', 'postmenopausal', 'peri', 'patients', 'individualized', 'homeopathic', 'trial', 'climacteric']\n",
      "['mdd', 'hippocampal volume', 'ngf', 'tpa', 'ect', 'patients', 'plasma bdnf', 'bdnf levels']\n",
      "['compulsive', 'knockout', 'glud1', 'reversal', 'behaviors', 'stn', 'ko', 'like', 'grooming', 'neurons']\n",
      "['paeoniflorin', 'tspo', 'postpartum depression', 'il', 'model', 'antidepressant', 'ppd', 'rats', 'postpartum']\n",
      "['successful treatment olfactory', 'olfactory reference syndrome', 'reference syndrome paroxetine', 'syndrome', 'olfactory', 'reference', 'case report', 'treatment olfactory', 'successful']\n",
      "['cbt', 'behavioral therapy', 'behavioral therapy sertraline', 'behavioral', 'childhood', 'sertraline childhood', 'combination cognitive', 'therapy', 'therapy sertraline', 'sertraline childhood anxiety']\n"
     ]
    }
   ],
   "source": [
    "params = \"alpha0.3_k10_res0.002\"\n",
    "words_to_remove = [\n",
    "    \"depression\",\n",
    "    \"fluoxetine\",\n",
    "    \"effects\",\n",
    "    \"effect\",\n",
    "    \"escitalopram\",\n",
    "    \"paroxetine\",\n",
    "    \"fluvoxamine\",\n",
    "    \"citalopram\",\n",
    "]\n",
    "\n",
    "df_dict = process_given_labels(graph_summary_df_dict, params, words_to_remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This code chunk performs community detection on a graph using different parameter configurations. It iterates over a list of parameter strings, extracts the necessary values from each string, and creates a partition based on the graph and the extracted parameters. It then drops unnecessary columns from the partition's dataframe, adds cluster labels to the graph, and saves the resulting graph as a GraphML file. Finally, it saves the partition's dataframe as a pickle file.\n",
    "\"\"\"\n",
    "\n",
    "final_params_list = [\n",
    "    \"alpha0.3_k10_res0.002\",\n",
    "    \"alpha0.3_k20_res0.01\",\n",
    "    \"alpha0.3_k20_res0.006\",\n",
    "]\n",
    "iterations = 50\n",
    "\n",
    "for params in final_params_list:\n",
    "    G = params_graph_dict[params.split(\"_res\")[0]]\n",
    "    resolution = pd.to_numeric(params.split(\"res\")[1])\n",
    "    column_name = f\"cluster_{params}\"\n",
    "    pc = PartitionCreator(G, df)\n",
    "    pc.create_partition_from_cmpvertexpartition(\n",
    "        n_iterations=iterations,\n",
    "        resolution_parameter=resolution,\n",
    "        verbose=False,\n",
    "        cluster_colunm_name=column_name,\n",
    "        centrality_column_name=f\"centrality_{params}\",\n",
    "    )\n",
    "    path = f\"../data/06-clustered-df/{params}.pkl\"\n",
    "    # drop all columns starting with 'centrality' that dont contain params\n",
    "    cols_to_drop = [\n",
    "        col for col in pc.df.columns if \"centrality\" in col or \"cluster\" in col\n",
    "    ]\n",
    "    cols_to_drop = [col for col in cols_to_drop if params not in final_params_list]\n",
    "    pc.df.drop(cols_to_drop, axis=1, inplace=True)\n",
    "    # add cluster label to graph\n",
    "    eid_cluster_dict = dict(zip(pc.df[\"eid\"], pc.df[f\"cluster_{params}\"]))\n",
    "    eid_centrality_dict = dict(zip(pc.df[\"eid\"], pc.df[f\"centrality_{params}\"]))\n",
    "    G.vs[\"cluster\"] = [eid_cluster_dict[eid] for eid in G.vs[\"eid\"]]\n",
    "    G.vs[\"cluster_ev_centrality\"] = [eid_centrality_dict[eid] for eid in G.vs[\"eid\"]]\n",
    "    # save graphml\n",
    "    G.write_graphml(f\"../data/07-clustered-graphs/{params}.graphml\")\n",
    "\n",
    "pc.df.to_pickle(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "\n",
    "Gpajek = G.copy()\n",
    "\n",
    "# Remove all attributes except 'eid' from each node\n",
    "for node, data in Gpajek.nodes(data=True):\n",
    "    keys_to_remove = [k for k in data if k != \"eid\"]  # Collect keys to remove\n",
    "    for k in keys_to_remove:\n",
    "        data.pop(k, None)  # Remove the key\n",
    "\n",
    "# Now, convert all remaining node attributes to strings\n",
    "for node, data in Gpajek.nodes(data=True):\n",
    "    for k, v in data.items():\n",
    "        data[k] = str(v)  # Convert attribute to string\n",
    "\n",
    "# Gpajek is now modified with only 'eid' attributes retained and converted to string\n",
    "# You can now save Gpajek to the Pajek format\n",
    "nx.write_pajek(Gpajek, \"f../data/07-clustered-graphs/alpha0.3_k20_res0.006.net\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NOW WE FOLLOW WITH MAIN PATH\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
